### PR TITLE
Migrate AccessControl to component

### LINKED ISSUE #769: Migrate AccessControl to component
**🧐 Motivation**

**📝 Details**

### PR DESCRIPTION
**Pull request by ericnordelo**
Fixes #769  

#### PR Checklist

- [x] Tests
- [ ] Tried the feature on a public network
- [ ] Documentation

### CODE REVIEW

File: src/tests/mocks/accesscontrol_mocks.cairo [hunk 1]
```diff
@@ -0,0 +1,209 @@
...
+
+    #[abi(embed_v0)]
+    impl AccessControlImpl =
+        accesscontrol_component::AccessControlImpl<ContractState>;
+    #[abi(embed_v0)]
+    impl AccessControlCamelImpl =
+        accesscontrol_component::AccessControlCamelImpl<ContractState>;
+    #[abi(embed_v0)]
+    impl SRC5Impl = src5_component::SRC5Impl<ContractState>;
+
```
Comment (andrew-fleming) [inline src/tests/mocks/accesscontrol_mocks.cairo:19]: Should we have `SRC5CamelImpl` here also?

File: src/tests/mocks/accesscontrol_mocks.cairo [hunk 2]
```diff
@@ -0,0 +1,209 @@
...
+    use starknet::ContractAddress;
+
+    component!(path: accesscontrol_component, storage: accesscontrol, event: AccessControlEvent);
+    component!(path: src5_component, storage: src5, event: SRC5Event);
+
+    #[abi(embed_v0)]
+    impl AccessControlCamelImpl =
+        accesscontrol_component::AccessControlCamelImpl<ContractState>;
+    #[abi(embed_v0)]
+    impl SRC5Impl = src5_component::SRC5Impl<ContractState>;
```
Comment (andrew-fleming) [inline src/tests/mocks/accesscontrol_mocks.cairo:97]: Same question: shouldn't this be camel?

File: src/tests/utils.cairo [hunk 3]
```diff
@@ -1,6 +1,7 @@
 mod constants;
 
 use starknet::ContractAddress;
+use starknet::class_hash::Felt252TryIntoClassHash;
```
Comment (andrew-fleming) [inline src/tests/utils.cairo:4]:
```suggestion
```
This import really doesn't want to leave

File: src/access/accesscontrol/accesscontrol.cairo [hunk 4]
```diff
@@ -72,92 +66,173 @@ mod AccessControl {
...
+        ) {
+            let admin = self.get_role_admin(role);
             self.assert_only_role(admin);
             self._revoke_role(role, account);
         }
 
-        fn renounce_role(ref self: ContractState, role: felt252, account: ContractAddress) {
+        /// Revokes `role` from the calling account.
+        ///
+        /// Roles are often managed via `grantRole` and `revokeRole`: this function's
```
Comment (andrew-fleming) [inline src/access/accesscontrol/accesscontrol.cairo:120]:
```suggestion
        /// Roles are often managed via `grant_role` and `revoke_role`: this function's
```

**Review by andrew-fleming — COMMENTED** [review]
Looking good, Eric! I left some comments and suggestions, and we need to update the Access API and doc pages

File: src/access/accesscontrol/accesscontrol.cairo [hunk 5]
```diff
@@ -72,92 +66,173 @@ mod AccessControl {
...
+        +HasComponent<TContractState>,
+        +SRC5::HasComponent<TContractState>,
+        +Drop<TContractState>
+    > of InternalTrait<TContractState> {
+        /// Initializes the contract by registering the IAccessControl interface Id.
+        fn initializer(ref self: ComponentState<TContractState>) {
+            let mut contract = self.get_contract_mut();
+            let mut src5_component = SRC5::HasComponent::<
+                TContractState
+            >::get_component_mut(ref contract);
```
[the remark this answers is not in the document]
Comment (ericnordelo) [reply src/access/accesscontrol/accesscontrol.cairo:188]: Cairo team said they were working on improving this, so probably worth waiting a bit anyway, but I agree.

File: src/tests/mocks/accesscontrol_mocks.cairo (hunk 1 above)
Comment (ericnordelo) [reply src/tests/mocks/accesscontrol_mocks.cairo:19]: We are removing the support for SRC5Camel, not on this PRs, but we shouldn't add more dependencies IMO.

Comment (andrew-fleming) [reply src/tests/mocks/accesscontrol_mocks.cairo:19]: I was thinking that, but it looks...wrong haha as long as we get rid rid of it before the next release

**Review by andrew-fleming — APPROVED** [review]
LGTM! Let's remember to update the AccessControl code examples in the ERC20 migration

**Pull request merged**